ios/RNGestureHandlerManager.m

```diff
@@ -132,25 +128,35 @@ - (void)handleClearJSResponder
 
 #pragma mark Root Views Management
 
-- (void)registerRootViewIfNeeded:(UIView*)childView
+- (void)registerRootViewGestureRecognizerIfNeeded:(UIView*)childView
 {
     UIView *parent = childView;
-    while (parent != nil && ![parent isKindOfClass:[RCTRootView class]]) parent = parent.superview;
-    
-    RCTRootView *rootView = (RCTRootView *)parent;
-    UIView *rootContentView = rootView.contentView;
-    if (rootContentView != nil && ![_rootViews containsObject:rootContentView]) {
-        RCTLifecycleLog(@"[GESTURE HANDLER] Initialize gesture handler for root view %@", rootContentView);
-        [_rootViews addObject:rootContentView];
-        RNRootViewGestureRecognizer *recognizer = [RNRootViewGestureRecognizer new];
-        recognizer.delegate = self;
-        rootContentView.userInteractionEnabled = YES;
-        [rootContentView addGestureRecognizer:recognizer];
+    while (parent != nil && ![parent respondsToSelector:@selector(touchHandler)]) parent = parent.superview;
+
+    UIView *rootView = [[parent performSelector:@selector(touchHandler)] view];
+  
+    if (rootView == nil) {
+      return;
     }
+  
+    for (RNRootViewGestureRecognizer *recognizer in _rootViewGestureRecognizers) {
+      // Many views can return the same touchHandler so we check if the one we want to register
```

Comment:
This comment should be moved to the place where we extract `view` from `touchHandler` rather than here